Marlin/src/module/temperature.cpp

```diff
@@ -768,26 +1219,70 @@ volatile bool Temperature::raw_temps_ready = false;
     #endif
     #if HAS_HEATED_CHAMBER
-      case H_CHAMBER: return temp_chamber.soft_pwm_amount;
+      case H_CHAMBER: return heaters[HeatedChamberPos].info.soft_pwm_amount;
     #endif
     #if HAS_COOLER
-      case H_COOLER: return temp_cooler.soft_pwm_amount;
+      case H_COOLER: return heaters[CoolerPos].info.soft_pwm_amount;
     #endif
     default:
-      return TERN0(HAS_HOTEND, temp_hotend[heater_id].soft_pwm_amount);
+      return TERN0(HAS_HOTEND, heaters[HotEndPos0 + heater_id].info.soft_pwm_amount);
   }
 }
 
 #define _EFANOVERLAP(A,B) _FANOVERLAP(E##A,B)
 
 #if HAS_AUTO_FAN
 
-  #define CHAMBER_FAN_INDEX HOTENDS
+  // A structure that's describing the conditions required to
+  // update a fan automatically.
+  struct FanTemperature {
+    const heater_pos_t pos;
+    const celsius_t    temp;
+    const bool         bit;
+    const pin_t        pin;
+
+    void autoUpdateFan(uint8_t speed) {
+      if (PWM_PIN(pin) && speed < 255) analogWrite(pin, bit ? speed : 0);
+      else WRITE(pin, bit);
+    }
+    virtual int autoUpdateFan() { autoUpdateFan(EXTRUDER_AUTO_FAN_SPEED); return 1; }
+
+    void check() { bit |= heaters[pos].deg() >= temp; }
+
+    FanTemperature(const heater_pos_t pos, const celsius_t temp, const pin_t pin)
+      : pos(pos), temp(temp), bit(false), pin(pin) {}
+  };
+  template <bool autoFan, int LimitTemperature, uint8_t fanSpeed = EXTRUDER_AUTO_FAN_SPEED>
+  struct FanTemperatureImpl : FanTemperature {
+    virtual int autoUpdateFan() { if (autoFan) { autoUpdateFan(fanSpeed); return 1; } return 0; }
+    FanTemperatureImpl(const heater_pos_t pos, const pin_t pin) : FanTemperature(pos, celsius_t(LimitTemperature, true), pin) {}
+  };
+
+  static FanTemperature fanTemperatures[] = {
+    #if HAS_HOTEND
+      #define DeclareFHotEnd(N) FanTemperatures<TERN0(HAS_AUTO_FAN_## N, true), celsius_t(EXTRUDER_AUTO_FAN_TEMPERATURE).raw()>{HotEndPos## N, E## N ##_AUTO_FAN_PIN },
+      LIST_BY_HOTENDS(DeclareFHotEnd(0), DeclareFHotEnd(1), DeclareFHotEnd(2), DeclareFHotEnd(3), DeclareFHotEnd(4), DeclareFHotEnd(5), DeclareFHotEnd(6), DeclareFHotEnd(7)),
+      #undef DeclareFHotEnd
+    #endif
+
+    #if HAS_AUTO_CHAMBER_FAN
+      FanTemperatures<TERN0(HAS_AUTO_CHAMBER_FAN && !AUTO_CHAMBER_IS_E, true),  celsius_t(CHAMBER_AUTO_FAN_TEMPERATURE).raw(), CHAMBER_AUTO_FAN_SPEED>{HeatedChamberPos, CHAMBER_AUTO_FAN_PIN },
+    #endif
+
+    #if HAS_AUTO_COOLER_FAN
+      FanTemperatures<true,  celsius_t(COOLER_AUTO_FAN_TEMPERATURE).raw(), 255>{CoolerPos, EO_AUTO_FAN_PIN }, // Not sure about the pin here
+    #endif
+  };
 
   void Temperature::checkExtruderAutoFans() {
+
+    // I don't understand what this does, and I don't think it's required anymore
+    // If it's used to map an index to a fan pin (and there are multiple index leading to the same pin),
+    // then instead, we should store a bit index in the FanTemperature's above
+    /*
     #define _EFAN(B,A) _EFANOVERLAP(A,B) ? B :
```

Comment:
@thinkyhead If you have some time, can you explain what this does ?
As far as I have understood, each heater can have a fan assigned to it, but different heater can have the same fan, right ?
So this is used to ensure we don't turn on a fan twice uselessly ?
How do you map from Heater to Fan index (I'm a bit lost with this macro here) ?

Reply:
```cpp
#define _FANOVERLAP(A,B) (A##_AUTO_FAN_PIN == E##B##_AUTO_FAN_PIN)
#define _EFANOVERLAP(A,B) _FANOVERLAP(E##A,B)
::
#if HAS_AUTO_FAN_1 && !_EFANOVERLAP(1,0) . . .
#if HAS_AUTO_FAN_1 && !_FANOVERLAP(E1,0) . . .
#if HAS_AUTO_FAN_1 && !(E1_AUTO_FAN_PIN == E0_AUTO_FAN_PIN)
```